File: generator/src/test/resources/kotlin/expected/ResponseEventsTest/test-event-stream-method-generation-for-common-base-events.output.kt
package io.test.service

import io.outfoxx.sunday.MediaType
import io.outfoxx.sunday.RequestFactory
import io.outfoxx.sunday.http.Method
import io.test.Base
import io.test.Test1
import io.test.Test2
import kotlin.collections.List
import kotlin.reflect.typeOf
import kotlinx.coroutines.flow.Flow

public class API(
  public val requestFactory: RequestFactory,
  public val defaultContentTypes: List<MediaType> = listOf(),
  public val defaultAcceptTypes: List<MediaType> = listOf(),
) {
  public fun fetchEventsSimple(): Flow<Base> = this.requestFactory
    .eventStream(
      method = Method.Get,
      pathTemplate = "/test1",
      acceptTypes = listOf(MediaType.EventStream),
      decoder = { decoder, _, _, data, _ -> decoder.decode<Base>(data, typeOf<Base>()) }
    )

  public fun fetchEventsDiscriminated(): Flow<Base> = this.requestFactory
    .eventStream(
      method = Method.Get,
      pathTemplate = "/test2",
      acceptTypes = listOf(MediaType.EventStream),
      decoder = { decoder, event, _, data, logger ->
        when (event) {
          "Test1" -> decoder.decode<Test1>(data, typeOf<Test1>())
          "Test2" -> decoder.decode<Test2>(data, typeOf<Test2>())
          else -> {
            logger.error("Unknown event type, ignoring event: event=$event")
            null
          }
        }
      }
    )
}

File: generator/src/test/resources/swift/expected/ResponseEventsTest/test-event-stream-method-generation-for-common-base-events.output.swift
import Sunday

public final class API<TransportType : Transport> : Sendable {

  public static var problemTypes: [ProblemRegistration] {
    return []
  }
  public let transport: TransportType
  public let defaultContentTypes: [MediaType]
  public let defaultAcceptTypes: [MediaType]

  public init(
    transport: TransportType,
    defaultContentTypes: [MediaType] = [],
    defaultAcceptTypes: [MediaType] = [],
    problemTypes: [ProblemRegistration] = API.problemTypes
  ) {
    self.transport = transport
    self.defaultContentTypes = defaultContentTypes
    self.defaultAcceptTypes = defaultAcceptTypes
    problemTypes.forEach { $0.register(on: transport) }
  }

  public func fetchEventsSimple() -> AsyncStream<any Base> {
    return self.transport.eventStream(
      method: .get,
      pathTemplate: "/test1",
      pathParameters: nil,
      queryParameters: nil,
      body: Empty.none,
      contentTypes: nil,
      acceptTypes: [.eventStream],
      headers: nil,
      decoder: { decoder, _, _, data, _ in try decoder.decode(BaseRef.self, from: data).value }
    )
  }

  public func fetchEventsDiscriminated() -> AsyncStream<any Base> {
    return self.transport.eventStream(
      method: .get,
      pathTemplate: "/test2",
      pathParameters: nil,
      queryParameters: nil,
      body: Empty.none,
      contentTypes: nil,
      acceptTypes: [.eventStream],
      headers: nil,
      decoder: { decoder, event, _, data, log in
        switch event {
        case "Test1": return try decoder.decode(Test1.self, from: data)
        case "Test2": return try decoder.decode(Test2.self, from: data)
        default:
          log.error("Unknown event type, ignoring event: event=\(event ?? "<none>", privacy: .public)")
          return nil
        }
      }
    )
  }

}
